File: idakonyvtar/src/main/java/eu/eyan/idakonyvtar/controller/LibraryController.java
package eu.eyan.idakonyvtar.controller;

import static eu.eyan.idakonyvtar.controller.input.BookControllerInput.ISBN_ENABLED;
import static javax.swing.JFileChooser.APPROVE_OPTION;

import java.awt.Component;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.io.File;

import javax.swing.JFileChooser;
import javax.swing.JFrame;
import javax.swing.JMenuBar;
import javax.swing.JOptionPane;
import javax.swing.JToolBar;
import javax.swing.ListSelectionModel;
import javax.swing.SwingUtilities;
import javax.swing.event.ListSelectionEvent;
import javax.swing.event.ListSelectionListener;
import javax.swing.filechooser.FileNameExtensionFilter;

import com.jgoodies.binding.adapter.SingleListSelectionAdapter;

import eu.eyan.idakonyvtar.controller.adapter.LibraryListTableModel;
import eu.eyan.idakonyvtar.controller.input.BookControllerInput;
import eu.eyan.idakonyvtar.controller.input.LibraryControllerInput;
import eu.eyan.idakonyvtar.model.Book;
import eu.eyan.idakonyvtar.model.LibraryModel;
import eu.eyan.idakonyvtar.util.DialogHelper;
import eu.eyan.idakonyvtar.util.ExcelHandler;
import eu.eyan.idakonyvtar.util.HighlightRenderer;
import eu.eyan.idakonyvtar.util.OkCancelDialog;
import eu.eyan.idakonyvtar.util.SpecialCharacterRowFilter;
import eu.eyan.idakonyvtar.view.LibraryMenuAndToolBar;
import eu.eyan.idakonyvtar.view.LibraryView;

public class LibraryController implements IControllerWithMenu<LibraryControllerInput, Void>, ActionListener
{

    public static final String NO = "Nem";
    public static final String YES = "Igen";
    public static final String TITLE = "IdaKönyvtár";
    private final LibraryMenuAndToolBar menuAndToolBar = new LibraryMenuAndToolBar();
    private final LibraryView view = new LibraryView();
    private final LibraryModel model = new LibraryModel();
    private Book previousBook;
    HighlightRenderer highlightRenderer = new HighlightRenderer();

    // FIXME: initview and getview are not the same!!
    @Override
    public Component getView()
    {
        view.getComponent();
        resetTableModel();
        view.getBookTable().setSelectionModel(new SingleListSelectionAdapter(model.getBooks().getSelectionIndexHolder()));
        view.getBookTable().setEnabled(true);
        view.getBookTable().setDefaultRenderer(Object.class, highlightRenderer);
        view.getBookTable().setSelectionMode(ListSelectionModel.SINGLE_SELECTION);
        return view.getComponent();
    }

    private void resetTableModel()
    {
        // FIXME: is it really neccessary????
        if (model.getBooks().getSize() > 0)
        {
            view.getBookTable().setEmptyText("Ilyen szűrőfeltételekkel nem található book.");
        }
        else
        {
            view.getBookTable().setEmptyText("Nincs book a listában.");
        }
        LibraryListTableModel dataModel = new LibraryListTableModel(model.getBooks(), model.getLibrary().getColumns(), model.getLibrary().getConfiguration());
        view.getBookTable().setModel(dataModel);
    }

    @Override
    public String getTitle()
    {
        return TITLE;
    }

    @Override
    public void initData(LibraryControllerInput input)
    {
        readLibrary(input.getFile());
        previousBook = new Book(model.getLibrary().getColumns().size());
    }

    private void readLibrary(File file)
    {
        System.out.println("Loading file: " + file);
        model.setLibrary(ExcelHandler.readLibrary(file));
        model.getBooks().getList().clear();
        // FIXME heee? 2x in model
        model.getBooks().setList(model.getLibrary().getBooks());
        resetTableModel();
    }

    private void saveLibrary(File file)
    {
        ExcelHandler.saveLibrary(file, model.getLibrary());
    }

    @Override
    public void initBindings()
    {
        menuAndToolBar.MENU_EXCEL_LOAD.addActionListener(this);
        menuAndToolBar.MENU_EXCEL_SAVE.addActionListener(this);

        menuAndToolBar.TOOLBAR_NEW_BOOK.addActionListener(this);
        menuAndToolBar.TOOLBAR_BOOK_DELETE.addActionListener(this);

        view.getBookTable().getSelectionModel().addListSelectionListener(new ListSelectionListener()
        {
            @Override
            public void valueChanged(ListSelectionEvent e)
            {
                menuAndToolBar.TOOLBAR_BOOK_DELETE.setEnabled(view.getBookTable().getSelectedRow() >= 0);
            }
        });

        view.getBookTable().addMouseListener(new MouseAdapter()
        {
            @Override
            public void mouseClicked(MouseEvent e)
            {
                if (e.getClickCount() == 2)
                {
                    editBook();
                }
            }

        });

        menuAndToolBar.TOOLBAR_SEARCH.addKeyListener(new KeyAdapter()
        {
            @Override
            public void keyReleased(KeyEvent e)
            {
                view.getBookTable().setRowFilter(new SpecialCharacterRowFilter(menuAndToolBar.TOOLBAR_SEARCH.getText()));
                highlightRenderer.setHighlightText(menuAndToolBar.TOOLBAR_SEARCH.getText());
            }
        });

        JFrame frame = (JFrame) SwingUtilities.getRoot(view.getComponent());
        frame.setDefaultCloseOperation(JFrame.DO_NOTHING_ON_CLOSE);
        frame.addWindowListener(new WindowAdapter()
        {
            @Override
            public void windowClosing(WindowEvent e)
            {
                if (DialogHelper.yesNo(frame, "Biztos ki akar lépni?", "Megerősítés"))
                {
                    frame.dispose();
                }
            }
        });
    }

    private void editBook()
    {
        BookController bookController = new BookController();
        int selectedBookIndex = view.getBookTable().convertRowIndexToModel(view.getBookTable().getSelectedRow());
        OkCancelDialog editorDialog = DialogHelper.startModalDialog(
                view.getComponent(),
                bookController,
                new BookControllerInput.Builder()
                        .withBook(new Book(model.getBooks().getList().get(selectedBookIndex)))
                        .withColumns(model.getLibrary().getColumns())
                        .withColumnConfiguration(model.getLibrary().getConfiguration())
                        .withBookList(model.getLibrary().getBooks())
                        .build());
        if (editorDialog.isOk())
        {
            model.getBooks().getList().set(selectedBookIndex, bookController.getOutput());
            // TODO: ugly: use selectioninlist...
            model.getBooks().fireSelectedContentsChanged();
        }
    }

    @Override
    public JMenuBar getMenuBar()
    {
        return menuAndToolBar.getMenuBar();
    }

    @Override
    public void actionPerformed(final ActionEvent e)
    {
        if (e.getSource() == menuAndToolBar.MENU_EXCEL_LOAD)
        {
            JFileChooser jFileChooser = new JFileChooser(".");
            jFileChooser.setApproveButtonText("Töltés");
            jFileChooser.setFileFilter(new FileNameExtensionFilter("Excel97 fájlok", "xls"));
            if (jFileChooser.showOpenDialog(menuAndToolBar.MENU_EXCEL_LOAD) == APPROVE_OPTION)
            {
                readLibrary(jFileChooser.getSelectedFile());
            }
        }

        if (e.getSource() == menuAndToolBar.MENU_EXCEL_SAVE)
        {
            JFileChooser jFileChooser = new JFileChooser(new File("."));
            jFileChooser.setApproveButtonText("Mentés");
            jFileChooser.setFileFilter(new FileNameExtensionFilter("Excel97 fájlok", "xls"));
            if (jFileChooser.showOpenDialog(menuAndToolBar.MENU_EXCEL_SAVE) == APPROVE_OPTION)
            {
                System.out.println("Save " + jFileChooser.getSelectedFile());
                saveLibrary(jFileChooser.getSelectedFile());
            }
        }

        if (e.getSource() == menuAndToolBar.TOOLBAR_NEW_BOOK)
        {
            BookController bookController = new BookController();

            OkCancelDialog editorDialog = DialogHelper.startModalDialog(
                    view.getComponent()
                    , bookController
                    , new BookControllerInput.Builder()
                            .withBook(newPreviousBook(model.getLibrary().getColumns().size()))
                            .withBookList(model.getBooks().getList())
                            .withColumns(model.getLibrary().getColumns())
                            .withIsbnEnabled(ISBN_ENABLED)
                            .withColumnConfiguration(model.getLibrary().getConfiguration())
                            .build());
            if (editorDialog.isOk())
            {
                model.getBooks().getList().add(0, bookController.getOutput());
                savePreviousBook(bookController.getOutput());
                // TODO: ugly: use selectioninlist...
                model.getBooks().fireIntervalAdded(0, 0);
            }
        }

        if (e.getSource() == menuAndToolBar.TOOLBAR_BOOK_DELETE)
        {
            if (JOptionPane.showOptionDialog(menuAndToolBar.TOOLBAR_BOOK_DELETE, "Biztosan törölni akarod?", "Törlés megerősítése", JOptionPane.YES_NO_OPTION, JOptionPane.QUESTION_MESSAGE, null, new String[] { YES, NO }, NO) == JOptionPane.OK_OPTION)
            {
                int selectionIndex = model.getBooks().getSelectionIndex();
                model.getBooks().getList().remove(selectionIndex);
                // TODO: ugly: use selectioninlist...
                model.getBooks().fireIntervalRemoved(selectionIndex, selectionIndex);
            }
        }
    }

    private void savePreviousBook(Book book)
    {
        previousBook = new Book(model.getLibrary().getColumns().size());
        for (String rememberingColumn : model.getLibrary().getConfiguration().getRememberingColumns())
        {
            int columnIndex = model.getLibrary().getColumns().indexOf(rememberingColumn);
            previousBook.setValue(columnIndex, book.getValue(columnIndex));
        }
    }

    private Book newPreviousBook(int size)
    {
        Book newBook = new Book(size);
        for (String rememberingColumn : model.getLibrary().getConfiguration().getRememberingColumns())
        {
            int columnIndex = model.getLibrary().getColumns().indexOf(rememberingColumn);
            newBook.setValue(columnIndex, previousBook.getValue(columnIndex));
        }
        return newBook;
    }

    @Override
    public Void getOutput()
    {
        return null;
    }

    @Override
    public JToolBar getToolBar()
    {
        return menuAndToolBar.getToolBar();
    }

    @Override
    public Component getComponentForFocus()
    {
        return menuAndToolBar.TOOLBAR_SEARCH;
    }
}

File: idakonyvtar/src/main/scala/eu/eyan/idakonyvtar/controller/LibraryController.scala
package eu.eyan.idakonyvtar.controller

import java.awt.Component
import java.io.File

import scala.collection.JavaConversions.asScalaBuffer

import com.jgoodies.binding.adapter.SingleListSelectionAdapter
import com.jgoodies.binding.list.SelectionInList

import eu.eyan.idakonyvtar.controller.adapter.LibraryListTableModel
import eu.eyan.idakonyvtar.controller.input.BookControllerInput
import eu.eyan.idakonyvtar.model.Book
import eu.eyan.idakonyvtar.model.ColumnConfigurations
import eu.eyan.idakonyvtar.model.Library
import eu.eyan.idakonyvtar.text.TextsIda
import eu.eyan.idakonyvtar.util.DialogHelper
import eu.eyan.idakonyvtar.util.ExcelHandler
import eu.eyan.idakonyvtar.util.LibraryException
import eu.eyan.idakonyvtar.view.BookTable
import eu.eyan.log.Log
import eu.eyan.util.io.FilePlus.FilePlusImplicit
import eu.eyan.util.jgoodies.SelectionInListPlus.SelectionInListImplicit
import eu.eyan.util.rx.lang.scala.subjects.BehaviorSubjectPlus.BehaviorSubjectImplicit
import eu.eyan.util.string.StringPlus.StringPlusImplicit
import eu.eyan.util.swing.HighlightRenderer
import eu.eyan.util.swing.JPanelWithFrameLayout
import eu.eyan.util.swing.JTablePlus.JTableImplicit
import eu.eyan.util.swing.SpecialCharacterRowFilter
import eu.eyan.util.swing.SwingPlus.showErrorDialog
import eu.eyan.util.swing.WithComponent
import javax.imageio.ImageIO
import javax.swing.JOptionPane
import javax.swing.ListSelectionModel
import rx.lang.scala.subjects.BehaviorSubject
import eu.eyan.util.scala.Try
import eu.eyan.util.scala.TryCatch
import eu.eyan.util.scala.TryCatchThrowable
import eu.eyan.idakonyvtar.text.TechnicalTextsIda._
import javax.swing.JScrollPane

class LibraryController(val file: File) extends WithComponent {
  override def toString = s"LibraryController[file=$file, nrOfBooks=${numberOfBooks.get}, isBookSelected=${isBookSelected.get}]"
  def getComponent = component
  private val bookTable = new BookTable(file.getPath)
  private val highlightRenderer = new HighlightRenderer
  private val books: SelectionInList[Book] = new SelectionInList[Book]()
  private val texts = new TextsIda

  val isBookSelected = BehaviorSubject(false)
  val numberOfBooks = BehaviorSubject(books.getList.size)
  books.onListData(numberOfBooks.onNext(books.getList.size))
  val isDirty = BehaviorSubject(false)
  def dirty = isDirty.onNext(true)
  def notDirty = isDirty.onNext(false)

  val component = new JScrollPane(new JPanelWithFrameLayout()
    .withBorders
    .withSeparators
    .newColumnFPG
    .newRowFPG
    .addInScrollPane(bookTable))

  Log.info("Loading file: " + file)

  private val library = Try(ExcelHandler.readLibrary(file))

  books.getList.clear
  library.foreach(lib => books.setList(lib.books))
  resetTableModel

  val previousBook = library.map(lib => Book(lib.columns.size)).getOrElse(Book(0))
  resetTableModel
  bookTable.setSelectionModel(new SingleListSelectionAdapter(books.getSelectionIndexHolder))
  bookTable.setEnabled(true)
  bookTable.setDefaultRenderer(classOf[Object], highlightRenderer)
  bookTable.setSelectionMode(ListSelectionModel.SINGLE_SELECTION)

  bookTable.onDoubleClick(editBook)
  bookTable.onValueChanged(isBookSelected.onNext(bookTable.getSelectedRow >= 0))

  def filter(textToFilter: String) = {
    bookTable.setRowFilter(new SpecialCharacterRowFilter(textToFilter))
    highlightRenderer.setHighlightText(textToFilter)
  }

  def deleteBook(parent: Component) = {
    if (JOptionPane.OK_OPTION == JOptionPane.showOptionDialog( // TODO DialogHelper.yesNo
      parent,
      "Biztosan törölni akarod?",
      "Törlés megerősítése",
      JOptionPane.YES_NO_OPTION,
      JOptionPane.QUESTION_MESSAGE,
      null,
      Array("Igen", "Nem"),
      "Nem")) {

      val selectionIndex = books.getSelectionIndex
      books.getList.remove(selectionIndex)
      books.fireIntervalRemoved(selectionIndex, selectionIndex)
      dirty
    }
  }

  private def resetTableModel() = {
    if (library.isFailure)
      bookTable.setEmptyText(ERROR_AT_READING_LIBRARY + ": " + library.failed.get.getMessage)
    else if (books.getSize > 0)
      bookTable.setEmptyText("Ilyen szűrőfeltételekkel nem található book.")
    else
      bookTable.setEmptyText("Nincs book a listában.")

    library.foreach(library => bookTable.setModel(LibraryListTableModel(books, library.columns.toList, library.configuration)))
  }

  def saveLibrary = {
    Log.info("Save " + file)
    library.foreach(library => Try { ExcelHandler.saveLibrary(file, library); notDirty }) // TODO: error alert if failed.
  }

  def saveAsLibrary(newFile: File) = {
    Log.info("SaveAs " + file)
    def confirmOverwrite = DialogHelper.yesNo(null, texts.SaveAsOverwriteConfirmText(newFile), texts.SaveAsOverwriteConfirmWindowTitle, texts.SaveAsOverwriteYes, texts.SaveAsOverwriteNo)
    library.foreach(library => Try { if (newFile.notExists || confirmOverwrite) ExcelHandler.saveLibrary(newFile, library) })
  }

  def createNewBook = library.foreach(library => {
    val bookControllerInput = new BookControllerInput(newPreviousBook(library.columns.size), library.columns.toList, library.configuration, books.getList.toList, true, file)
    val bookController = new BookController

    val editorDialog = DialogHelper.startModalDialog(component, bookController, bookControllerInput)

    if (editorDialog.isOk) {
      saveImages(bookController.getOutput)
      books.getList.add(0, bookController.getOutput)
      savePreviousBook(bookController.getOutput)
      books.fireIntervalAdded(0, 0)
      dirty
    }
  })

  def editBook = library.foreach(library => {
    val selectedBookIndex = bookTable.convertRowIndexToModel(bookTable.getSelectedRow)

    val bookControllerInput = new BookControllerInput(Book(books.getList.get(selectedBookIndex)), library.columns.toList, library.configuration, library.books.toList, false, file)
    val bookController = new BookController

    val editorDialog = DialogHelper.startModalDialog(component, bookController, bookControllerInput)

    if (editorDialog.isOk) {
      saveImages(bookController.getOutput)
      books.getList.set(selectedBookIndex, bookController.getOutput)
      books.fireSelectedContentsChanged
      dirty
    }
  })

  private def saveImages(book: Book) = library.foreach(library => {
    //TODO WTF is this spagetti? put all relevant shot to book and only tha save belongs here.
    val columns = library.columns.toList
    val columnConfiguration = library.configuration
    for { i <- 0 until columns.size } {
      val imageName = book.getValue(i)
      val columnName = columns(i)
      val isPictureField = columnConfiguration.isTrue(columnName, ColumnConfigurations.PICTURE)
      if (imageName == "" && book.images.contains(i)) {
        val dir = file.getParentFile
        val imagesDir = (file.getAbsolutePath + ".images").asDir
        if (!imagesDir.exists) imagesDir.mkdir
        val imageFile = (imagesDir.getAbsolutePath + "\\IMG.JPG").asFile.generateNewNameIfExists()
        ImageIO.write(book.images.get(i).get, "JPG", imageFile)
        book.values(i) = imageFile.getName
      } else {
        // do nothing image already there
      }
    }
  })

  private def savePreviousBook(book: Book) = library.foreach(library => {
    library.configuration.getRememberingColumns.foreach(colName => {
      val columnIndex = library.columns.indexOf(colName)
      Log.info("LibraryController.savePreviousBook")
      previousBook.setValue(columnIndex, book.getValue(columnIndex))
    })
  })

  private def newPreviousBook(size: Int): Book = {
    val newBook = Book(size)

    library.foreach(library => {
      library.configuration.getRememberingColumns.foreach(rememberingColumn => {
        val columnIndex = library.columns.indexOf(rememberingColumn)
        newBook.setValue(columnIndex, previousBook.getValue(columnIndex))
        Log.info("LibraryController.newPreviousBook remembering col " + columnIndex + " val:" + previousBook.getValue(columnIndex))
      })
    })

    newBook
  }
}